File: Exp1/Python/src/Vector_x_Matrix.py
import time
import numpy as np


def v_x_m(vct, mtx):
    n = 4096
    ans = [0] * n
    for j in range(n):
        for i in range(n):
            ans[j] += vct[i] * mtx[i][j]


with open('vector.data', 'r') as f:
    string = f.readline().split()
    vector = [int(s) for s in string]

with open('matrix.data', 'r') as f:
    strings = f.readlines()
    matrix = []
    for ss in strings:
        matrix.append([int(s) for s in ss.split()])

v = np.array(vector)
m = np.array(matrix)

start_time = time.time_ns()
v_x_m(vector, matrix)
# v.dot(m)
end_time = time.time_ns()
round_time = (end_time - start_time) / (10 ** 6)

print('Python time', round(round_time, 2), 'ms')

File: Exp1/C/src/Vector_x_Matrix.c
#include <stdio.h>
#include <time.h>

#define N 4096

void v_x_m(int v[N], int m[N][N]) {
	int c[N] = { 0 };
	for (int j = 0; j < N; j++) {
		for (int i = 0; i < N; i++) {
			c[i] += v[j] * m[i][j];
		}
	}
}

int vct[N] = { 0 }, mtx[N][N] = { 0 };

int main() {
	freopen("vector.data", "r", stdin);
	for (int i = 0; i < N; i++) {
		scanf("%d", &vct[i]);
	}
	fclose(stdin);

	freopen("matrix.data", "r", stdin);
	for (int i = 0; i < N; i++) {
		for (int j = 0; j < N; j++) {
			scanf("%d", &mtx[i][j]);
		}
	}
	fclose(stdin);
	
	
	struct timespec time_start = { 0, 0 }, time_end = { 0, 0 };
	clock_gettime(CLOCK_REALTIME, &time_start);
	v_x_m(vct, mtx);
	clock_gettime(CLOCK_REALTIME, &time_end);
	long long int roundtime = (time_end.tv_sec - time_start.tv_sec) * 1000000000 + time_end.tv_nsec - time_start.tv_nsec;

	double t = roundtime / 1000000.0;
	printf("C time %.2f ms\n", t);

	return 0;
}
